File: middleware-jackson/src/main/java/com/nytimes/android/external/store3/middleware/jackson/JacksonBufferedSourceAdapter.java
package com.nytimes.android.external.store3.middleware.jackson;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.nytimes.android.external.fs3.BufferedSourceAdapter;

import java.io.ByteArrayInputStream;

import javax.annotation.Nonnull;
import javax.inject.Inject;

import okio.BufferedSource;
import okio.Okio;

/**
 * An implementation of {@link BufferedSourceAdapter} that uses {@link ObjectMapper} to convert Java values to JSON.
 */
public class JacksonBufferedSourceAdapter<Parsed> implements BufferedSourceAdapter<Parsed> {

    private final ObjectMapper objectMapper;

    @Inject
    public JacksonBufferedSourceAdapter(@Nonnull ObjectMapper objectMapper) {
        this.objectMapper = objectMapper;
    }

    @Nonnull
    @Override
    public BufferedSource toJson(@Nonnull Parsed value) {
        try {
            return Okio.buffer(Okio.source(new ByteArrayInputStream(objectMapper.writeValueAsBytes(value))));
        } catch (JsonProcessingException e) {
            throw new RuntimeException(e);
        }
    }
}

File: middleware-jackson/src/main/java/com/nytimes/android/external/store3/middleware/jackson/JacksonBufferedSourceAdapter.kt
package com.nytimes.android.external.store3.middleware.jackson

import com.fasterxml.jackson.databind.ObjectMapper
import com.nytimes.android.external.fs3.BufferedSourceAdapter
import okio.BufferedSource
import okio.Okio
import java.io.ByteArrayInputStream
import javax.inject.Inject

/**
 * An implementation of [BufferedSourceAdapter] that uses [ObjectMapper] to convert Java values to JSON.
 */
class JacksonBufferedSourceAdapter<Parsed> @Inject
constructor(private val objectMapper: ObjectMapper) : BufferedSourceAdapter<Parsed> {
    override fun toJson(value: Parsed): BufferedSource = Okio.buffer(Okio.source(ByteArrayInputStream(objectMapper.writeValueAsBytes(value))))
}
